osquery/remote/http/http_client.cpp

```diff
@@ -0,0 +1,342 @@
+      : pc_(pc) {}
+
+  const char* name() const noexcept {
+    return pc_->name();
+  }
+
+  std::string message(int ev) const {
+    return pc_->message(ev);
+  }
+
+ private:
+  boost::system::error_category const* pc_;
+};
+
+/** In the postResponseHandler, treating SHORT_READ_ERROR as success
+ *  for ssl connections. This can happen if a remote server did not
+ *  call shutdown on ssl connection.
+ */
+void Client::postResponseHandler(boost_system::error_code const& ec) {
+  if ((ec.category() == boost_asio::error::ssl_category) &&
+      (ec.value() == SHORT_READ_ERROR)) {
+    // ignoring short read error
+    ec_ = boost_system::errc::make_error_code(boost_system::errc::success);
+  } else if ((ec.value() != boost_system::errc::operation_canceled) ||
+             (ec.category() != boost_asio::error::system_category)) {
+    ec_ = ec;
+  }
+}
+
+void Client::closeSocket() {
+  if (sock_.is_open()) {
+    boost_system::error_code rc;
+    sock_.shutdown(boost_asio::ip::tcp::socket::shutdown_both, rc);
+    if (rc) {
+      return;
+    }
+    sock_.close(rc);
+  }
+}
+
+void Client::timeoutHandler(boost_system::error_code const& ec) {
+  if (!ec) {
+    closeSocket();
+    ec_ = boost_system::errc::make_error_code(boost_system::errc::timed_out);
+  }
+}
+
+void Client::createConnection() {
+  if (!client_options_.remote_hostname_) {
+    throw std::runtime_error("Remote hostname missing");
+  }
+
+  std::string port = (client_options_.proxy_hostname_)
+                         ? std::to_string(PROXY_DEFAUT_PORT)
+                         : *client_options_.remote_port_;
+
+  std::string connect_host = (client_options_.proxy_hostname_)
+                                 ? *client_options_.proxy_hostname_
+                                 : *client_options_.remote_hostname_;
+
+  std::size_t pos;
+  if ((pos = connect_host.find(":")) != std::string::npos) {
+    port = connect_host.substr(pos + 1);
+    connect_host = connect_host.substr(0, pos);
+  }
+
+  closeSocket();
+  boost_system::error_code rc;
+  connect(sock_,
+          r_.resolve(boost_asio::ip::tcp::resolver::query{connect_host, port}),
+          rc);
+
+  if (rc) {
+    std::string error("Failed to connect to ");
+    if (client_options_.proxy_hostname_) {
+      error += "proxy host ";
+    }
+    error += connect_host + ":" + port;
+    throw std::system_error(
+        std::error_code(rc.value(), std::generic_category()), error);
+  }
+
+  if (client_options_.proxy_hostname_) {
+    std::string remote_host = *client_options_.remote_hostname_;
+    std::string remote_port = *client_options_.remote_port_;
+
+    beast_http_request req;
+    req.method(beast_http::verb::connect);
+    req.target(remote_host + ":" + remote_port);
+    req.version = 11;
+    req.prepare_payload();
+    beast_http::write(sock_, req);
+
+    boost::beast::flat_buffer b;
+    beast_http_response_parser rp;
+    rp.skip(true);
+    beast_http::read_header(sock_, b, rp);
```

Comment:
What if the response contains "Content-Length: 1000"? This code will happily accept it, but in fact it is an error according to rfc7230. Consider using `beast_http::response<beast_http::empty_body>` instead, and call `read` instead of `read_header`.  The parser will generate an error if the message indicates there is any payload.

Reply:
This code part it contacting the proxy server. Even if we receive the 1000 bytes we still want to connect to proxy.

Reply:
It depends on how strict you want your implementation to adhere to rfc7230. If the proxy server is not under your control you might want to be more strict.